segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorage.java

```diff
@@ -0,0 +1,201 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.bookkeeperstorage;
+
+import io.pravega.common.util.ImmutableDate;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Getter;
+import org.apache.curator.framework.api.transaction.CuratorOp;
+
+/*
+ * Storage Log represents a single segment. A segment is rolled over whenever a ownership change is observed.
+ *
+ **/
+class LogStorage {
+
+    private final LogStorageManager manager;
+
+    @Getter
+    private final String name;
+
+    @Getter
+    private final boolean readOnlyHandle;
+
+    private final ConcurrentHashMap<Integer, LedgerData> dataMap;
```

Comment:
I suggest making this non-concurrent and guarding it with the same lock as the other fields. I see in various places that you either modify it under the main lock or not. Let's not use double-locking.